oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/client/BanyanDBSchemaMapper.java

```diff
@@ -0,0 +1,119 @@
+     */
+    private static final Map<String, String> KEY_MAPPING = new HashMap<>();
+
+    static {
+        // Known mapping
+        // 1) duration -> latency
+        // 2) state -> is_error
+        // others:
+        // http.method, status_code, db.type, db.instance, mq.queue, mq.broker, mq.topic
+        KEY_MAPPING.put("duration", SegmentRecord.LATENCY);
+        KEY_MAPPING.put("state", SegmentRecord.IS_ERROR);
+        KEY_MAPPING.put("http.method", TAG_PREFIX + "http.method");
+        KEY_MAPPING.put("status_code", TAG_PREFIX + "status_code");
+        KEY_MAPPING.put("db.type", TAG_PREFIX + "db.type");
+        KEY_MAPPING.put("db.instance", TAG_PREFIX + "db.instance");
+        KEY_MAPPING.put("mq.queue", TAG_PREFIX + "mq.queue");
+        KEY_MAPPING.put("mq.broker", TAG_PREFIX + "mq.broker");
+        KEY_MAPPING.put("mq.topic", TAG_PREFIX + "mq.topic");
+    }
+
+    /**
+     * Map the key defined in BanyanDB schema to the Factory Lambda which can be invoked to build the Field
+     */
+    private final Map<String, Function<Map<String, Object>, Write.Field>> fieldsMap = new HashMap<>();
+
+    public BanyanDBSchemaMapper(Set<String> schemaKeys) {
+        this.schemaKeys = schemaKeys;
+
+        // iterate over keys in the schema
+        for (final String key : schemaKeys) {
+            fieldsMap.put(key, entityMap -> buildField(entityMap.get(KEY_MAPPING.getOrDefault(key, key))));
+        }
+    }
+
+    @Override
+    public Write.EntityValue apply(SegmentRecord segmentRecord) {
+        final Map<String, Object> objectMap = entity2Storage(segmentRecord);
+
+        return Write.EntityValue.newBuilder()
+                .addAllFields(this.schemaKeys.stream().map(s -> fieldsMap.get(s).apply(objectMap)).collect(Collectors.toList()))
+                .setDataBinary(ByteString.copyFrom(segmentRecord.getDataBinary()))
+                .setTimestamp(Timestamp.newBuilder().setSeconds(segmentRecord.getTimeBucket()).build())
+                .setEntityId(segmentRecord.getSegmentId()).build();
+    }
+
+    static Write.Field buildField(Object value) {
+        if (value == null) {
+            return Write.Field.newBuilder().setNull(NullValue.NULL_VALUE).build();
+        }
+        if (value instanceof String) {
+            return Write.Field.newBuilder().setStr(Write.Str.newBuilder().setValue((String) value).build()).build();
+        } else if (value instanceof Integer) {
+            return Write.Field.newBuilder().setInt(Write.Int.newBuilder().setValue((Integer) value).build()).build();
+        } else if (value instanceof Long) {
+            return Write.Field.newBuilder().setInt(Write.Int.newBuilder().setValue((Long) value).build()).build();
+        } else {
+            throw new IllegalStateException("should not reach here");
+        }
+    }
+
+    static Map<String, Object> entity2Storage(SegmentRecord segmentRecord) {
```

Comment:
I don't like this code style. You have all the things after `SegmentRecord#Builder#entity2Storage`. Why build another Map?

Reply:
If there is really a necessary to override this, you should do this through `StorageDAO#newRecordDao`. Builder has been given as the parameter(`StorageHashMapBuilder`). You could read generic type to get the real metrics/record type, and build a required DAO.

SkyWalking's modularization is powerful to override. Don't do hijack and hardcode.

Reply:
The basic ideas here are,

1. We do not need to encode/decode binary in Base64 format.
2. Some of the tags have to be treated as indexed field as they are defined explicity in the schema.

Reply:
If you take a look at what I mentioned, all these could be done there.